### PR TITLE
Mecha pilots!

### PR DESCRIPTION
**Pull request by RemieRichards**
The Syndicate have trained a new division of fighters, no longer mindless grunts who wish to shoot their friends or run up to you with an esword, these fighters pilot Mecha! Which means they do all of the above again, but deadlier!
- Added Mecha Pilots, Hostile mobs that attempt to find and pilot mecha. REAL MECHA!
- Cleaned out some ancient Mecha shitcode
- Generalised all Mecha powers down to the basic `/mecha` level, to allow easier addition of those powers/abilities to other Mecha, and to allow the Syndie Pilot AI to be simpler
- Gives hostile mobs a `targets_from` var, the same treatment I did for turrets before, allowing them to use their AI from within "vehicles" or other mobs
- Adds a `PossibleThreats()` proc which returns a list of mobs just as the normal hostile AI flow would, allowing you to get a list of things this mob would attack (useful for finding the number of enemies around your hostile mob)
- Adds a var to allow mob AI to walk/move inside of other atoms 

WeaponsTest

The scecond gif doesn't work and I don't know why so I cri
PunchyPunchy

### CODE REVIEW

**Comment by lzimann** [comment]
This can be great for away missions

**Comment by Zilenan91** [comment]
So NPC mechs?

**Comment by PKPenguin321** [comment]
so is this NPC mechs or NPCs that pilot mechs?

**Comment by RemieRichards** [comment]
@Zilenan91, no, those are real mecha, piloted by a mob AI, if you break the mecha or it runs out of charge the mob bails out and looks for another one to steal.

**Comment by Zilenan91** [comment]
Are the mobs always DNA locked to the mechs or can you steal them if you EMP them enough.

**Comment by PKPenguin321** [comment]
that's so cool

**Comment by RemieRichards** [comment]
@Zilenan91 you can steal their mecha if they exit it, yes, Provided you have the access (Eg: if they're in a Syndie mecha you still need syndie access)

**Comment by Zilenan91** [comment]
Jeez, I hope mappers put in some failsafes for these if they want to put these in their away missions, letting players have maulers and dark gyaxes could be problematic.

**Comment by PKPenguin321** [comment]
it would be cool if some day we could port these to the SNPC system instead of having simple mobs

**Comment by RemieRichards** [comment]
@PKPenguin321 SNPCs are deprecated.

**Comment by PKPenguin321** [comment]
it's a shame, really

**Comment by KazeEspada** [comment]
If you manage to capture one of these mechs, you fucking deserve it.

**Comment by Zilenan91** [comment]
I guess yeah but all you'd need is an EMP grenade or two and the Ion rifle. Then just bonk the dude who gets out badda boom godmode achieved.

**Comment by PKPenguin321** [comment]
iirc the nuke op mechs are actually super fucking tanky and could probably definitely survive way more than 3 emps

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 1]
```diff
@@ -162,11 +189,13 @@
 				Goto(target,move_to_delay,minimum_distance)//Otherwise, get to our minimum distance so we chase them
 		else
 			Goto(target,move_to_delay,minimum_distance)
-		if(isturf(loc) && target.Adjacent(src))	//If they're next to us, attack
-			AttackingTarget()
-		return 1
+		if(target)
```
Comment (phil235) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: why the target check?

Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: I don't know, it's not necessary as it's checked above, probably a hold over from when I was debugging the AI.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm [hunk 2]
```diff
@@ -64,41 +72,58 @@
...
-	var/Target = PickTarget(Targets)
+	var/Target = PickTarget(.)
 	GiveTarget(Target)
 	return Target //We now have a target
 
+
+
+/mob/living/simple_animal/hostile/proc/PossibleThreats()
+	. = list()
+	for(var/pos_targ in ListTargets())
```
Comment (phil235) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: ListTargets() is a super costly proc, please try to call PossibleThreats() in life() and give the list of targets calculated in life() as an argument to that proc.

Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: That's, not doable, I need this list in automated action, and I can't risk having an out of date list for the cycle, so I can't store it on the mob either.

**Comment by drynwyn** [comment]
EMP's will destroy mecha before they drain their battery IIRC, so while you could blow one up it'd just be wreckage.

File: code/modules/mob/living/simple_animal/hostile/hostile.dm (hunk 1 above)
Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/hostile/hostile.dm]: >CANNOT EXECUTE NULL.ADJACENT()

screw you phillll it is needed.

**Comment by Zilenan91** [comment]
EMPs cna't deal the finishing blow to mechs anymore afaik.

File: code/game/mecha/combat/combat.dm [hunk 3]
```diff
@@ -2,9 +2,6 @@
 	force = 30
 	internal_damage_threshold = 50
 	damage_absorption = list("brute"=0.7,"fire"=1,"bullet"=0.7,"laser"=0.85,"energy"=1,"bomb"=0.8)
-	var/am = "d3c2fbcadca903a41161ccc9df9cf948"
```
Comment (neersighted) [inline code/game/mecha/combat/combat.dm]: Just curious -- what was this for?

Comment (RemieRichards) [inline code/game/mecha/combat/combat.dm]: Not used, unless you want to dig the history of the Mecha folder, we'll never know.

**Comment by Razharas** [comment]
FIne

**Pull request merged**